src/processor/FHIRProcessor.ts

```diff
@@ -13,22 +20,31 @@ export class FHIRProcessor {
     this.fhir = fhir;
   }
 
-  process(inputPath: string) {
+  process(
+    inputPath: string
+  ): (ExportableProfile | ExportableExtension | ExportableCodeSystem | ExportableInvariant)[] {
     const rawContent = JSON.parse(fs.readFileSync(inputPath, 'utf-8'));
 
     if (rawContent['resourceType'] === 'StructureDefinition') {
       // Profiles and Extensions are both made from StructureDefinitions
+      // Invariants may be contained within StructureDefinitions
       this.structureDefinitions.push(rawContent);
       if (rawContent.type === 'Extension') {
         logger.debug(`Processing contents of ${inputPath} as Extension.`);
-        return ExtensionProcessor.process(rawContent, this.fhir);
+        const extension = ExtensionProcessor.process(rawContent, this.fhir);
+        if (extension) {
+          return [extension, ...InvariantProcessor.process(rawContent)];
+        }
```

Comment:
This isn't exactly what I expected, but I'm not sure it's _wrong_ -- it's just not what I expected.  I think it depends on how you view the processors.  Does the name of the processor indicate what it _processes_ or what it _returns_?  I.e., does `ExtensionProcessor` (which extends `StructureDefinitionProcessor`) _process_ Extension SDs (and therefore should return whatever it finds) or does it _return_ Extensions (and therefore should return only `Extension`s)?

I was thinking that an `ExtensionProcessor` processes Extension SDs, and therefore, it would return all the things that can result from processing Extension SDs: `ExportableExtension`, `ExportableInvariant`s, (eventually) `ExportableMapping`s, etc.  So rather than having a separate `InvariantProcessor`, we would add an `extractInvariants` function to the `StructureDefinitionProcessor` and its `process` function would now call that as well, to collect all the invariants -- and eventually return them along with the extension.

This is really a design question about what classes do what things where - and will establish a pattern for future features -- so it's probably worth some thought and discussion.  Glad to hear input from @mint-thompson, @ngfreiter, and @jafeltra on this one.

Reply:
Upon further reflection, I think I like your pattern of "A processor is named for what it takes as input." So invariants would get extracted, not processed (since they're not a top-level entity in the JSON). Handling them before any rules lets `InvariantExtractor` set the `processedPaths` appropriately to avoid extra caret value rules, since `ObeysRuleExtractor` doesn't care what paths have already been processed, and there's no harm in adding a duplicate entry to `processedPaths`.

A processor is named for its input and may have various outputs (Profile, Invariants, etc.) based on what exactly is found in the input. An extractor is named for its output and may have various inputs based on what exactly is needed (element, SD, FHIR defs, etc.) to generate the outputs. I feel like these can be consistent rules.

Relatedly, I think that the two separate `ProfileProcessor` and `ExtensionProcessor` can probably be safely refactored and removed in favor of a concrete `StructureDefinitionProcessor`. But, that's a completely separate task from this one.

Reply:
I think that all makes sense to me, and you've explained it in a way that I think we can generally apply in the future.